tests/unit/autojac/test_mtl_backward.py

```diff
@@ -0,0 +1,268 @@
+    """
+
+    p0 = torch.tensor([1.0, 2.0], requires_grad=True)
+    p1 = torch.tensor([1.0, 2.0], requires_grad=True)
+    p2 = torch.tensor([3.0, 4.0], requires_grad=True)
+
+    r1 = torch.tensor([-1.0, 1.0]) @ p0
+    r2 = (p0**2).sum() + p0.norm()
+    y1 = r1 * p1[0] + r2 * p1[1]
+    y2 = r1 * p2[0] + r2 * p2[1]
+
+    mtl_backward(
+        losses=[y1, y2],
+        features=[r1, r2],
+        tasks_params=[[], []],
+        shared_params=[],
+        A=UPGrad(),
+    )
+
+    for p in [p0, p1, p2]:
+        assert p.grad is None
+
+
+def test_mtl_backward_partial_parameters():
+    """
+    Tests that mtl_backward fills the right .grad values when only a subset of the parameters
+    are specified as inputs.
+    """
+
+    p0 = torch.tensor([1.0, 2.0], requires_grad=True)
+    p1 = torch.tensor([1.0, 2.0], requires_grad=True)
+    p2 = torch.tensor([3.0, 4.0], requires_grad=True)
+
+    r1 = torch.tensor([-1.0, 1.0]) @ p0
+    r2 = (p0**2).sum() + p0.norm()
+    y1 = r1 * p1[0] + r2 * p1[1]
+    y2 = r1 * p2[0] + r2 * p2[1]
+
+    mtl_backward(
+        losses=[y1, y2],
+        features=[r1, r2],
+        tasks_params=[[p1], []],
+        shared_params=[p0],
+        A=Mean(),
+    )
+
+    assert (p0.grad is not None) and (p0.shape == p0.grad.shape)
+    assert (p1.grad is not None) and (p1.shape == p1.grad.shape)
+    assert p2.grad is None
+
+
+def test_mtl_backward_empty_tasks():
+    """
+    Tests that mtl_backward raises an error when called with an empty list of tasks.
+    """
+
+    p0 = torch.tensor([1.0, 2.0], requires_grad=True)
+
+    r1 = torch.tensor([-1.0, 1.0]) @ p0
+    r2 = (p0**2).sum() + p0.norm()
+
+    with pytest.raises(ValueError):
+        mtl_backward(
+            losses=[],
+            features=[r1, r2],
+            tasks_params=[],
+            shared_params=[p0],
+            A=UPGrad(),
+        )
+
+
+def test_mtl_backward_incoherent_task_number():
+    """
+    Tests that mtl_backward raises an error when called with the number of tasks losses
+    different from the number of tasks parameters.
+    """
+
+    p0 = torch.tensor([1.0, 2.0], requires_grad=True)
+    p1 = torch.tensor([1.0, 2.0], requires_grad=True)
+    p2 = torch.tensor([3.0, 4.0], requires_grad=True)
+
+    r1 = torch.tensor([-1.0, 1.0]) @ p0
+    r2 = (p0**2).sum() + p0.norm()
+    y1 = r1 * p1[0] + r2 * p1[1]
+    y2 = r1 * p2[0] + r2 * p2[1]
+
+    with pytest.raises(ValueError):
+        mtl_backward(
+            losses=[y1, y2],
+            features=[r1, r2],
+            tasks_params=[[p1]],  # Wrong
+            shared_params=[p0],
+            A=UPGrad(),
+            retain_graph=True,
+        )
+    with pytest.raises(ValueError):
+        mtl_backward(
+            losses=[y1],  # Wrong
+            features=[r1, r2],
+            tasks_params=[[p1], [p2]],
+            shared_params=[p0],
+            A=UPGrad(),
+            retain_graph=True,
```

Comment:
Here we could have in principle the default `False` value for `retain_graph` but it doesn't matter much.